### PR TITLE
Order finding quantum algorithm

### PR DESCRIPTION
**Pull request by msoeken**
This is an implementation of the order-finding algorithm described in the paper by L.M.K. Vandersypen, M. Steffen, G. Breyta, C.S. Yannoni, R. Cleve, and I.L. Chuang in [Experimental realization of an order-finding algorithm with an NMR quantum computer](https://doi.org/10.1103/PhysRevLett.85.5452), *Phys. Rev. Lett.* **85**, 5452, 2000.

It needs to realize a permutation with different exponents in the quantum circuit. The algorithm makes use of the `PermutationOracle` from the `ReversibleLogicSynthesis` project for this purpose. It also uses `QFT` from the canon.

The quantum algorithm is implemented for general permutation sizes (2ⁿ elements), however, the classical post-processing in the C# file is restricted to permutations over 4 elements.

I needed to make a small update to `PermutationOracle` to allow it to be used as a controlled operation.

### CODE REVIEW

File: Samples/src/OrderFinding/Program.cs [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+using Microsoft.Quantum.Simulation.Simulators;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.Quantum.Samples.OrderFinding
+{
+    /// <summary>
+    /// This class holds a permutation π and allows to compute its
+    /// order in various different ways (exact, guess classically,
```
Comment (martinquantum) [inline Samples/src/OrderFinding/Program.cs]:
```suggestion
    /// order in various different ways (exactly, classical guess, 
```

File: Samples/src/OrderFinding/Program.cs [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    /// guess with quantum)
```
Comment (martinquantum) [inline Samples/src/OrderFinding/Program.cs]:
```suggestion
    /// and quantum guess)
```

File: Samples/src/OrderFinding/Program.cs [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+            while (index != _p[cur])
+            {
+                ++order;
+                cur = _p[cur];
+            }
+            return order;
+        }
+
+        /// <summary>
+        /// Guesses the order (classically) for cycle that contains index
```
Comment (martinquantum) [inline Samples/src/OrderFinding/Program.cs]:
```suggestion
        /// Guesses the order (classically) for cycle that contains a given index
```

File: Samples/src/OrderFinding/Program.cs [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+        /// <param name="shots">Number of repetitions when guessing</param>
+        /// </summary>
+        public Permutation(List<int> permutation, int shots = 1024)
+        {
+            _p = permutation;
+            _shots = shots;
+        }
+
+        /// <summary>
+        /// Returns the exact order (length) of the cycle that contains given index
```
Comment (martinquantum) [inline Samples/src/OrderFinding/Program.cs]:
```suggestion
        /// Returns the exact order (length) of the cycle that contains a given index
```

File: Samples/src/OrderFinding/Program.cs [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+                return _rnd.Next(2) == 0 ? 1 : 3;
+            }
+            else
+            {
+                return _rnd.Next(2) == 0 ? 2 : 4;
+            }
+        }
+
+        /// <summary>
+        /// Guesses order classically for shots times, and returns the percentage for each order that was returned.
```
Comment (martinquantum) [inline Samples/src/OrderFinding/Program.cs]:
```suggestion
        /// Guesses the order classically by applying estimate `shots` many times and returning the percentage for each order that was returned.
```

File: Samples/src/OrderFinding/Program.cs [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+        /// find a period in the resulting state.  The result needs to be post-processed to find the estimate.
+        /// <summary>
+        private int GuessOrderQuantumOne(int index)
+        {
+            var result = OrderFinding.Run(_sim, new QArray<long>(_p.ConvertAll(x => (long)x)), (long)index).Result;
+
+            if (result == 0)
+            {
+                var guess = _rnd.NextDouble();
+                if (guess <= 0.5505)
```
Comment (martinquantum) [inline Samples/src/OrderFinding/Program.cs]: please explain magic constants for the branchings. Perhaps put a pointer to the original paper.

File: Samples/src/OrderFinding/Program.cs [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+                {
+                    return 1;
+                }
+                else if (guess <= 0.5505 + 0.1009)
```
Comment (martinquantum) [inline Samples/src/OrderFinding/Program.cs]: same here

File: Samples/src/OrderFinding/Program.cs [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+                {
+                    return 2;
+                }
+                else if (guess <= 0.5505 + 0.1009 + 0.1468)
```
Comment (martinquantum) [inline Samples/src/OrderFinding/Program.cs]: same here

File: Samples/src/OrderFinding/OrderFinding.qs [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+namespace Microsoft.Quantum.Samples.OrderFinding {
+    open Microsoft.Quantum.Extensions.Math;
+    open Microsoft.Quantum.Primitive;
+    open Microsoft.Quantum.Canon;
+    open Microsoft.Quantum.Samples.ReversibleLogicSynthesis;
+
+    /// # Summary
+    /// Given a permutation π, this function returns the
+    /// square of the permutation π².
```
Comment (martinquantum) [inline Samples/src/OrderFinding/OrderFinding.qs]:
```suggestion
    /// square of π, i.e., the permutation π².
```

File: Samples/src/OrderFinding/OrderFinding.qs [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+    function Square(perm : Int[]) : Int[] {
+        mutable sq_perm = new Int[Length(perm)];
+        for (i in 0..Length(perm) - 1) {
+            set sq_perm[i] = perm[perm[i]];
+        }
+        return sq_perm;
+    }
+
+    /// # Summary
+    /// Order Finding algorithm as described in [1]
```
Comment (martinquantum) [inline Samples/src/OrderFinding/OrderFinding.qs]:
```suggestion
    /// Implementes the Order Finding algorithm as described in L.M.K. Vandersypen et al., PRL 85, 5452, 2000 (https://arxiv.org/abs/quant-ph/0007017).
```

File: Samples/src/OrderFinding/OrderFinding.qs [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+    /// Finally, a QFT is applied to the upper qubits.
+    ///
+    /// # Input
+    /// ## perm
+    /// The input permutation
+    ///
+    /// ## index
+    /// Index of permutation
+    ///
+    /// [1] L.M.K. Vandersypen et al., PRL 85, 5452, 2000 (https://arxiv.org/abs/quant-ph/0007017)
```
Comment (martinquantum) [inline Samples/src/OrderFinding/OrderFinding.qs]:
```suggestion
```

**Comment by martinquantum** [comment]
Thanks for your contribution, Mathias! Please take a look at the suggested changes.

File: Samples/src/OrderFinding/Program.cs (hunk 6 above)
Comment (msoeken) [reply Samples/src/OrderFinding/Program.cs]: Fixed in 3e0ecdd.

File: Samples/src/OrderFinding/Program.cs (hunk 7 above)
Comment (msoeken) [reply Samples/src/OrderFinding/Program.cs]: Fixed in 3e0ecdd.

File: Samples/src/OrderFinding/Program.cs (hunk 8 above)
Comment (msoeken) [reply Samples/src/OrderFinding/Program.cs]: Fixed in 3e0ecdd.

**Comment by msoeken** [comment]
Thanks for your feedback, Martin! I merged your changes and added some explanation from where to read the probabilities in the `result == 0` case.

**Pull request merged**